File: src/2015/01/day.01.ts
import fs from "fs";

import "@/utils/array";

enum Instructions {
  UP = "(",
  DOWN = ")",
}

function parseInput() {
  return fs
    .readFileSync("src/inputs/2015/01/day.01.input.txt")
    .toString()
    .split("")
    .map(parseInstruction);
}

function parseInstruction(value: string) {
  switch (value) {
    case Instructions.UP: {
      return 1;
    }

    case Instructions.DOWN: {
      return -1;
    }

    default: {
      throw new Error("Unable to parse instruction!");
    }
  }
}

export function part1() {
  const instructions = parseInput();

  return instructions.sum();
}

export function part2() {
  const instructions = parseInput();

  let floor = 0;

  for (let i = 0; i < instructions.length; i++) {
    floor += instructions[i];

    if (floor < 0) {
      return i + 1;
    }
  }
}

File: src/2015/01/day.01.py
from timeit import default_timer as timer


def parse_input():
    with open("src/inputs/2015/01/day.01.input.txt", "r") as file:
        return file.readline()


def part1():
    instructions = parse_input()

    current_floor = 0

    for character in instructions:
        match character:
            case '(':
                current_floor += 1
            case ')':
                current_floor -= 1

    return current_floor


def part2():
    instructions = parse_input()

    current_floor = 0

    for (index, character) in enumerate(instructions):
        match character:
            case '(':
                current_floor += 1
            case ')':
                if current_floor == 0:
                    return index + 1

                current_floor -= 1

    return -1


def timed(fn):
    start = timer()
    result = fn()

    return (timer() - start, result)


(p1_elapsed, p1_answer) = timed(part1)
print(f"Part 1: {p1_answer} ({p1_elapsed:.3f}s)")

(p2_elapsed, p2_answer) = timed(part2)
print(f"Part 2: {p2_answer} ({p2_elapsed:.3f}s)")
